lib/net.c

```diff
@@ -95,73 +95,86 @@ randomize_rss_key(int guarantee_random_entropy)
+ethertype_flow_add(struct gatekeeper_if *iface, uint16_t ether_type,
 	uint16_t queue_id)
 {
-	struct rte_eth_ethertype_filter filter = {
-		.ether_type = rte_cpu_to_le_16(ether_type),
-		.flags = 0,
-		.queue = queue_id,
+	struct rte_flow_attr attr = { .ingress = 1 };
+	struct rte_flow_action_queue queue = { .index = queue_id };
+	struct rte_flow_action action[] = {
+		{
+			.type = RTE_FLOW_ACTION_TYPE_QUEUE,
+			.conf = &queue,
+	       	},
+		{
+			.type = RTE_FLOW_ACTION_TYPE_END,
+		}
 	};
-	unsigned int i;
+	struct rte_flow_item_eth eth_spec = {
+		.type = rte_cpu_to_be_16(ether_type),
+	};
+	struct rte_flow_item_eth eth_mask = {
+		.type = 0xFFFF,
+	};
+	struct rte_flow_item pattern[] = {
+		{
+			.type = RTE_FLOW_ITEM_TYPE_ETH,
+			.spec = &eth_spec,
+			.mask = &eth_mask,
+		},
+		{
+			.type = RTE_FLOW_ITEM_TYPE_END,
+		},
+	};
+	struct rte_flow_error error;
+	struct rte_flow *flow;
 	int ret;
 
-	RTE_VERIFY(iface->hw_filter_eth);
-
-	for (i = 0; i < iface->num_ports; i++) {
-		uint16_t port_id = iface->ports[i];
-		ret = rte_eth_dev_filter_ctrl(port_id,
-			RTE_ETH_FILTER_ETHERTYPE,
-			RTE_ETH_FILTER_ADD,
-			&filter);
-		if (ret == -ENOTSUP) {
-			G_LOG(NOTICE,
-				"net: hardware doesn't support adding an EtherType filter for 0x%02hx on port %hhu\n",
-				ether_type, port_id);
-			ret = -1;
-			goto out;
-		} else if (ret == -ENODEV) {
-			G_LOG(NOTICE,
-				"net: port %hhu is invalid for adding an EtherType filter for 0x%02hx\n",
-				port_id, ether_type);
-			ret = -1;
-			goto out;
-		} else if (ret != 0) {
-			G_LOG(NOTICE,
-				"net: other errors that depend on the specific operations implementation on port %hhu for adding an EtherType filter for 0x%02hx\n",
-				port_id, ether_type);
-			ret = -1;
-			goto out;
-		}
+	if (!iface->rss) {
+		/*
+		 * If RSS is not supported, then data plane packets
+		 * could be assigned to RX queues that are serviced
+		 * by non-data plane blocks (e.g., LLS).
+		 */
+		G_LOG(NOTICE, "net: cannot use EtherType filters when RSS is not supported\n");
+		return -1;
 	}
 
-	ret = 0;
+	ret = rte_flow_validate(iface->id, &attr, pattern, action, &error);
+	if (ret < 0) {
+		/*
+		 * A negative errno value was returned
+		 * (and also put in rte_errno).
+		 */
+		G_LOG(NOTICE, "net: could not validate EtherType flow: %s (%s)\n",
+			strerror(-ret), error.message);
+		return -1;
+	}
 
-out:
-	return ret;
-}
+	flow = rte_flow_create(iface->id, &attr, pattern, action, &error);
+	if (flow == NULL) {
+		/* rte_errno is set to a positive errno value. */
+		G_LOG(NOTICE, "net: could not create EtherType flow: %s (%s)\n",
+			strerror(rte_errno), error.message);
+		return -1;
+	}
 
-/*
- * ntuple filters can only be used if supported by the NIC
- * (to steer matching packets) and if RSS is supported
- * (to steer non-matching packets elsewhere).
- */
-static inline bool
-hw_filter_ntuple_available(const struct gatekeeper_if *iface)
-{
-	return iface->hw_filter_ntuple && iface->rss;
+	return 0;
 }
```

Comment:
These lines belong to the next patch, namely, **net: migrate ntuple filters to IPv4 flows**.

Notice that `return 0;` belongs to `ethertype_flow_add()`, not `hw_filter_ntuple_available()`.